concordium-consensus/src/Concordium/KonsensusV1/Consensus.hs

```diff
@@ -31,4 +41,185 @@ class MonadMulticast m where
+        Nothing -> do
+            -- The transaction is new to us. Before adding it to the transaction table,
+            -- we verify it.
+            theTime <- utcTimeToTimestamp <$> currentTime
+            verRes <- runTransactionVerifierT (TVer.verify theTime bi) =<< getCtx
+            case verRes of
+                (TVer.Ok res) -> insertTransaction res
+                notAccepted -> return $! NotAdded notAccepted
+  where
+    -- Insert the transaction into the transaction table and pending transaction table.
+    insertTransaction okRes = do
+        added <- addTransaction 0 bi $! TVer.Ok okRes
+        if added
+            then do
+                addPendingTransaction Individual bi
+                return $! Added bi $! TVer.Ok okRes
+            else -- If the transaction was not added it means it contained an old nonce.
+                return ObsoleteNonce
+    -- Create a context suitable for verifying a transaction within a 'Individual' context.
+    getCtx = do
+        _ctxSkovData <- get
+        _ctxBs <- bpState <$> gets' _lastFinalized
+        chainParams <- Concordium.GlobalState.BlockState.getChainParameters _ctxBs
+        let _ctxMaxBlockEnergy = chainParams ^. cpConsensusParameters . cpBlockEnergyLimit
+        return $! Context{_ctxTransactionOrigin = Individual, ..}
+    -- 'TransactionHash' of the transaction we're processing.
+    txHash = getHash bi
+
+-- |Attempt to put the 'BlockItem's of a 'BakedBlock' into the tree state.
+-- Return 'True' of the transactions were added otherwise 'False'.
+--
+-- Post-condition: Only transactions that are deemed verifiable
+-- (i.e. the verification yields a 'TVer.OkResult' or a 'TVer.MaybeOkResult') up to the point where
+-- a transaction processing might fail are added to the tree state.
+processBlockItems ::
+    forall m pv.
+    ( MonadProtocolVersion m,
+      IsConsensusV1 pv,
+      MonadState (SkovData pv) m,
+      BlockStateQuery m,
+      TimeMonad m,
+      AccountNonceQuery m,
+      MPV m ~ pv,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState (MPV m)
+    ) =>
+    -- |The baked block
+    BakedBlock ->
+    -- |Pointer to the parent block.
+    BlockPointer pv ->
+    -- |Return 'True' only if all transactions were
+    -- successfully processed otherwise 'False'.
+    m Bool
+processBlockItems bb parentPointer = process True $! bbTransactions bb
+  where
+    -- Create a context suitable for verifying a transaction within a 'Block' context.
+    getCtx = do
+        _ctxSkovData <- get
+        let _ctxBs = bpState parentPointer
+        chainParams <- Concordium.GlobalState.BlockState.getChainParameters _ctxBs
+        let _ctxMaxBlockEnergy = chainParams ^. cpConsensusParameters . cpBlockEnergyLimit
+        return $! Context{_ctxTransactionOrigin = Block, ..}
+    theRound = bbRound bb
+    -- Process the vector of transactions recursively.
+    process :: Bool -> Vector.Vector BlockItem -> m Bool
```

Comment:
Why not fold over the vector itself? It seems like constructing a list (O(n)) just for iterating is wasteful since we can just fold over the vector itself?